chips/e310x/src/chip.rs

```diff
@@ -76,35 +75,32 @@ impl<A: 'static + Alarm<'static>> kernel::Chip for E310x<A> {
     }
 
     fn service_pending_interrupts(&self) {
-        let mut reenable_intr = FieldValue::<u32, mie::Register>::new(0, 0, 0);
-
         loop {
             let mip = CSR.mip.extract();
 
             if mip.is_set(mip::mtimer) {
                 unsafe {
                     timer::MACHINETIMER.handle_interrupt();
                 }
-                reenable_intr += mie::mtimer::SET;
             }
             if mip.is_set(mip::mext) {
                 unsafe {
                     Self::handle_plic_interrupts();
                 }
-                reenable_intr += mie::mext::SET;
             }
 
             if !mip.matches_any(mip::mext::SET + mip::mtimer::SET) {
                 break;
             }
         }
 
-        // re-enable any interrupt classes which we handled
-        CSR.mie.modify(reenable_intr);
+        // Re-enable all MIE interrupts that we care about. Since we looped
+        // until we handled them all, we can re-enable all of them.
+        CSR.mie.modify(mie::mext::SET + mie::mtimer::SET);
```

Comment:
This doesn't seem right as we disable the mtimer when we want to disable the alarm: https://github.com/tock/tock/blob/master/arch/rv32i/src/machine_timer.rs#L141

Reply:
But the e310 platform (hifive) isn't using that SchedulerTimer implementation. I agree they aren't compatible, but we can't have the scheduler timer changing interrupts for the only timer source we have.

Reply:
If someone adds a RISC-V board that does have other timers they are going to look at this code (and OT which is the same) as an example. I don't think we should be enabling timer interrupts if they weren't already enabled AND have code that disables them and expects them to stay disabled..

Ideally I think it would be best to record that we disabled them handling an interrupt and then only enable them if that's true.

Reply:
I actually think that is more confusing, since Tock's design is all interrupts are enabled at the NVIC/MIE/PLIC layer, but putting that aside, I don't actually know how to write that code in a way that is both correct and I would be ok with someone else copying.

Reply:
But this (always re-enabling mext and mtimer) is clearly incorrect -- what if the kernel disabled mtimer (for whatever reason)? When there is only a single interrupt source, the scheduler timer should not be disabling its interrupts; it goes through a virtualizer. The virtualizer *may* disable underlying interrupts if there are no outstanding alarms. Breaking abstraction boundaries (e.g., who owns mtimer) in random ways due to software engineering issues is bad.

Reply:
This is especially pernicious for RISC-V, as the mtimer interrupt will fire continually on <= rather than only fire once on ==.

Reply:
Right now in Tock the kernel cannot dynamically disable the mtimer interrupt (it could choose not to use it at all, and it can disable it when it is also pending, but the design as I understand it is all interrupts at the NVIC/PLIC/CLIC/CLINT/MIE layer are always enabled unless pending).

Reply:
Even if it is always enabled, blinding force enabling it still seems wrong. What if someone is trying to debug something and wants the timer disabled? This will force enable it after it's disabled. There could also be future code changes that allow it to be disabled in the future.

Reply:
> but the design as I understand it is all interrupts at the NVIC/PLIC/CLIC/CLINT/MIE layer are always enabled unless pending).

Nope. This would mean you cannot suppress interrupts and therefore wakeup. Super-bad.

Instead, the current design -- which is fundamentally flawed -- is that you can suppress interrupts from occurring/waking up the MCU but cannot suppress their handling. I.e, #1306  . https://github.com/tock/tock/blob/ad9387a577405675b044d5bde85badf0274995c8/arch/cortex-m/src/nvic.rs#L167 and https://github.com/tock/tock/blob/ad9387a577405675b044d5bde85badf0274995c8/arch/cortex-m/src/nvic.rs#L109 let you disable interrupts in the CortexM, NVIC, for example.

The CortexM semantics are a problem for all kinds of reasons. But it's even *more* of a problem is the semantics aren't even consistent across architectures!

Reply:
I agree with @bradjc on this: it is currently how we are treating interrupts. I agree with @phil-levis and @alistair23 that is not necessarily the _right_ way to be dealing with interrupts, but that _is_ the architecture we're using in Tock atm.

And generally, you _cannot_ currently disable interrupts from waking the CPU: `nvic::disable` is an instance method of a struct that requires `unsafe` to construct and is handed to **nobody** and `nvic::disable_all` is `unsafe`.

Reply:
I don't understand why it would be too hard to just record that we disabled the interrupt in interrupt handling and then only re-enable interrupts if that happened. What am I missing here?